src/apps/vpn/connectionmanager.cpp

```diff
@@ -0,0 +1,999 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "connectionmanager.h"
+
+#include "app.h"
+#include "appconstants.h"
+#include "apppermission.h"
+#include "captiveportal/captiveportal.h"
+#include "controller.h"
+#include "controllerimpl.h"
+#include "dnshelper.h"
+#include "feature.h"
+#include "frontend/navigator.h"
+#include "ipaddress.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "models/devicemodel.h"
+#include "models/keys.h"
+#include "models/server.h"
+#include "models/servercountrymodel.h"
+#include "mozillavpn.h"
+#include "networkrequest.h"
+#include "rfc/rfc1112.h"
+#include "rfc/rfc1918.h"
+#include "rfc/rfc4193.h"
+#include "rfc/rfc4291.h"
+#include "serveri18n.h"
+#include "serverlatency.h"
+#include "settingsholder.h"
+#include "tasks/controlleraction/taskcontrolleraction.h"
+#include "tasks/function/taskfunction.h"
+#include "tasks/heartbeat/taskheartbeat.h"
+#include "taskscheduler.h"
+#include "tutorial/tutorial.h"
+
+#if defined(MZ_LINUX)
+#  include "platforms/linux/linuxcontroller.h"
+#elif defined(MZ_MACOS) || defined(MZ_WINDOWS)
+#  include "localsocketcontroller.h"
+#elif defined(MZ_IOS)
+#  include "platforms/ios/ioscontroller.h"
+#elif defined(MZ_ANDROID)
+#  include "platforms/android/androidcontroller.h"
+#else
+#  include "platforms/dummy/dummycontroller.h"
+#endif
+
+constexpr const uint32_t TIMER_MSEC = 1000;
+
+// X connection retries.
+constexpr const int CONNECTION_MAX_RETRY = 9;
+
+constexpr const uint32_t CONFIRMING_TIMOUT_SEC = 10;
+constexpr const uint32_t HANDSHAKE_TIMEOUT_SEC = 15;
+
+#ifndef MZ_IOS
+// The Mullvad proxy services are located at internal IPv4 addresses in the
+// 10.124.0.0/20 address range, which is a subset of the 10.0.0.0/8 Class-A
+// private address range.
+constexpr const char* MULLVAD_PROXY_RANGE = "10.124.0.0";
+constexpr const int MULLVAD_PROXY_RANGE_LENGTH = 20;
+#endif
+
+namespace {
+Logger logger("Connection Manager");
+
+ConnectionManager::Reason stateToReason(ConnectionManager::State state) {
+  if (state == ConnectionManager::StateSwitching ||
+      state == ConnectionManager::StateSilentSwitching) {
+    return ConnectionManager::ReasonSwitching;
+  }
+
+  if (state == ConnectionManager::StateConfirming) {
+    return ConnectionManager::ReasonConfirming;
+  }
+
+  return ConnectionManager::ReasonNone;
+}
+}  // namespace
+
+ConnectionManager::ConnectionManager() {
+  MZ_COUNT_CTOR(ConnectionManager);
+
+  m_connectingTimer.setSingleShot(true);
+  m_handshakeTimer.setSingleShot(true);
+
+  connect(&m_timer, &QTimer::timeout, this, &ConnectionManager::timerTimeout);
+
+  connect(&m_connectingTimer, &QTimer::timeout, this, [this]() {
+    m_enableDisconnectInConfirming = true;
+    emit enableDisconnectInConfirmingChanged();
+  });
+
+  connect(&m_handshakeTimer, &QTimer::timeout, this,
+          &ConnectionManager::handshakeTimeout);
+
+  LogHandler::instance()->registerLogSerializer(this);
```

Comment:
do we also need to unregister the log serializer? This PR removes code doing this in controller.cpp.

Reply:
yes! good catch